Stop using Mocha

utilum opened the pull request:
Following discussion with @bogdanvlviv I'm breaking this up into smaller manageable more review-friendly PRs, that will be submitted one by one.

From the lowest hanging fruit to deep embeded mocking.

- [x] Remove unnecessary mock objects (created with by `.stubs`) 8c1e1726d, 2af162c46, d2d7296
Working on these tests it became evident that some calls to `Mocha#stub` and `Mocha#stubs` do not serve the test case in any way. The same is true of some calls to `returns` on such objects. Remove both.
- [x] Reduce mocking by testing for value instead of method call 331cca175
Another low hanging fruit. When we can test for value, we do not need to test the setter method was called on a mock.
- [x] Replace shallow mocks with Ruby classes f7bfb3db2
Some mocks are simple to replace with a minimal class that responds to the needed methods.
- [x] Replace `Mocha#stubs` with `Minitest#stub 837d60317, 7afdc6c1a, a94d6ae81
- [x] Use MethodCallAssertions instead of Mocha part 1 d0743d026, d89937505
Replace simple instances of `Mocha#expect` with `MethodCallAssertions`.
- [x] Use MethodCallAssertions instead of Mocha part 2 82e42c1bd
Replace instances of `Mocha#expects` with `MethodCallAssertsions` where expecations need to be added to withstand Minitest's more stringent mocking.
- [x] Stub deeper mocks with new `MethodCallAssertions`
Remaining calls to `Mocha#stubs` and `Mocha#expects` cannot be replicated with plain Minitest. Stub and mock them with Ruby.
- [x] Remove mocha from Gemfile [_Q.E.F._]

bogdanvlviv reviewed (commented):
Hi @utilum. Thanks for working on this.
I've done first look at the PR and have some concerns I've mentioned below.
Also, What do you think about splitting this PR to small PRs? It will be easier to review and catch some not explicit mistakes if exists.

utilum commented:
Thanks _a lot_ @bogdanvlviv. I was beginning to think of breaking it into smaller PRs too, wondering if it's too bulky to be reviewed. No problem with that. I'll start working on your specific comments first, and perhaps garner some more input by the time they are answered.

bogdanvlviv commented:
To be honest It took me about 3 hours to review 85% of PR. :sweat_smile:.
(I can't imagine how much time it took you to prepare this PR. Thanks you for that :bowing_man: ) 
I think if we go with small PR, everyone will be more confident in changes we merge and It can prevent merging not explicit mistakes.

utilum commented:
Yeah... took me a good few days. Sometimes that's what's needed, to dive into the code.
And I see what you mean. Thank you for taking the time. I had thought that breaking into a logical progression of commits, each passing tests and advancing us toward the goal would facilitate review, but I guess it did not have the desired effect.
So... hoow about I change the description of this PR into a TODO list for smaller PRs, submit those, and check items as they are completed? This would conserve your comments as well.

bogdanvlviv commented:
I would recommend don't submit all PRs at once, instead go one by one(I mean open new one if previous one was merged).
Would be great to see simple PR with https://github.com/rails/rails/pull/33162/files#diff-81e98371a331e686a926985a7b9e9d88 activerecord/test/support/method_stub.rb file. It might help concentrate more people on it and get  more suggestions with possible improvements to it.

> This would conserve your comments as well.

Please don't worry I'll be happy to review again.

utilum commented:
Started acting on your suggestion @bogdanvlviv. I'm leaving this PR here for now, as a reference, and also, should the smaller PRs all work out, we'd have an easy reference for what connects them all.

kaspth commented:
I was just looking into reviewing this the other day. Then I skimmed the diff and thought "nope! not today". So 👍 on the smaller PRs.

I remember originally we wanted to replace Mocha because it took 300 ms to load, which was a pain for the railties tests as they fork a new process for each test or so. Is that still relevant? 😊

utilum commented:
I am not sure how long it takes, and failed to find the discussion that preceded all those `# FIXME: stop using mocha` comments. Any pointers?
What _did_ motivate me is making the tests more precise, with Minitest's strict mocks and stubs. Avoiding such lines as `SomeClass.any_instance.expects(:some_method).with(:some_arg).never`, as facilitated by Mohca.

utilum commented on activerecord/test/support/method_stub.rb:
```diff
@@ -0,0 +1,43 @@
+module MethodStub
+  private
+
+    def with_stubbed_method(object, method_name, return_value = nil)
```
I'm not sure this would be accurate, what about when we pass `SomeObject.singleton_class` as the object argument?

bogdanvlviv replied:
Currently, I have not the right answer. We might discuss it later on new short Pull Request
as I mentioned:
> Would be great to see simple PR with https://github.com/rails/rails/pull/33162/files#diff-81e98371a331e686a926985a7b9e9d88 activerecord/test/support/method_stub.rb file. It might help concentrate more people on it and get more suggestions with possible improvements to it.

utilum commented on activerecord/test/support/method_stub.rb:
```diff
@@ -0,0 +1,43 @@
+
+        define_method("stubbed_#{method_name}") do |*args|
+          eval "@@#{base_var_name}_called += 1"
+          eval "@@#{base_var_name}_args += args"
```
How else will we define these class variables dynamically? With `send` ?

bogdanvlviv replied:
I guess using of just `@@#{base_var_name}_called += 1` should work as well.
And what about using of instance variabes instead of class variables then we are able to use `remove_instance_variable` http://ruby-doc.org/core-2.5.1/Object.html#method-i-remove_instance_variable

utilum replied:
You are right! Removing the redundant `eval`s.

kaspth commented:
@utilum thanks for all your work on this! I just merged step 6, was that all that's left?

utilum commented:
My pleasure, @kaspth , the process has been great. There are two steps left: handling deeper mocks that require `class_eval`, and removal of the obviated requirement and Gemfile entry. I'm preparing both.

utilum commented on activerecord/test/support/method_stub.rb:
```diff
@@ -0,0 +1,43 @@
+    def unstub_method(object, method_name)
+      object.class_eval(<<-EOF, __FILE__, __LINE__ + 1)
+        alias_method(method_name, "original_#{method_name}")
+        remove_method("stubbed_#{method_name}")
```
I'm adding `remove_method` as you suggest, and `remove_class_variable` for the variables.

utilum committed:
Stub with Minitest and test with MethodCallAssertions

utilum committed:
Add method_call_assertions and use them instead of Mocha

Six Mocha calls prove quite resistant to Minitestification. For example,
if we replace

```
  ActiveRecord::Associations::HasManyAssociation
    .any_instance
    .expects(:reader)
    .never
```

with `assert_not_called`, Minitest wisely raises

```
NameError: undefined method `reader' for class `ActiveRecord::Associations::HasManyAssociation'
```

as `:reader` comes from a deeply embedded abstract class,
`ActiveRecord::Associations::CollectionAssociation`.

This patch tackles this difficulty by adding
`ActiveSupport::Testing::MethodCallAsserts#assert_called_on_instance_of`
which injects a stubbed method into `klass`, and verifies the number of
times it is called, similar to `assert_called`. It also adds  a convenience
method, `assert_not_called_on_instance_of`, mirroring
`assert_not_called`.

It uses the new method_call_assertions to replace the remaining Mocha
calls in `ActiveRecord` tests.

[utilum + bogdanvlviv + kspath]

utilum committed:
Remove Mocha from Gemfile

_Q.E.F._

utilum commented:
@bogdanvlviv , @kaspth I think this is finally ready

bogdanvlviv commented on activerecord/test/support/method_stub.rb:
```diff
@@ -0,0 +1,53 @@
+    def with_stubbed_method(object, method_name, return_value = nil)
+      stub_method(object, method_name, return_value)
+
+      yield
```
what about something like
`with_stubbed_method(object, method_name, expect_called_times: nil, returns: nil)`
and after this `yield` check
```ruby
if expect_called_times
  assert_equal expect_called_times, object.class_variable_get("@@#{base_var_name}_called")
end
(untested code)

utilum replied:
I'm not sure... what is the benefit?

`with_stubbed_method` is used and named as a helper. If we add the assertion we'd also want to rename it something like `assert_called_with_stubbed_method`, which does not describe the cases where the assertion is not made. We can adjust your proposal to add the new method and 

And then, I like the we need to explicitly assert the expectation.

bogdanvlviv commented:
We also should add test cases for `#with_stubbed_method` in the similar way we've done for `assert_called` an its friends. See:`activesupport/test/testing/method_call_assertions_test.rb`

utilum commented:
Do we test other methods from `rails_component/test/support/` or `rails_component/test/helper.rb`?

bogdanvlviv commented on activerecord/test/support/method_stub.rb:
```diff
@@ -0,0 +1,53 @@
+    # TODO: `alias_method` is private in Ruby 2.4, so once the minimum
+    # version is 2.5 there'll be no need to resort to `module_eval`.
+
+    def stub_method(object, method_name, return_value = nil)
```
It seems works only if `object` is a Class. If pass an instance as `object` it doesn't work, does it?
I still think that it should be renamed to `with_stubbed_instance_method` since it only can stub method for any instance of the specific class.
For stubbing any method of an object we have `assert_called_with`
@utilum What do you think about it?

utilum replied:
Why not?

bogdanvlviv commented:
I've been playing with this a bit.
What do you think about adding `assert_not_called_instance_method` and `assert_called_instance_method`?

You can apply this diff (via `git apply`) if you agree.
 

```diff
diff --git a/activerecord/test/cases/associations/eager_test.rb b/activerecord/test/cases/associations/eager_test.rb
index d38c736fa5..622516f78f 100644
--- a/activerecord/test/cases/associations/eager_test.rb
+++ b/activerecord/test/cases/associations/eager_test.rb
@@ -1,7 +1,6 @@
 # frozen_string_literal: true
 
 require "cases/helper"
-require "support/method_stub"
 require "models/post"
 require "models/tagging"
 require "models/tag"
@@ -31,8 +30,6 @@
 require "models/contract"
 
 class EagerAssociationTest < ActiveRecord::TestCase
-  include MethodStub
-
   fixtures :posts, :comments, :authors, :essays, :author_addresses, :categories, :categories_posts,
             :companies, :accounts, :tags, :taggings, :people, :readers, :categorizations,
             :owners, :pets, :author_favorites, :jobs, :references, :subscribers, :subscriptions, :books,
@@ -1574,10 +1571,8 @@ def test_preloading_has_many_through_with_custom_scope
 
   # CollectionProxy#reader is expensive, so the preloader avoids calling it.
   test "preloading has_many_through association avoids calling association.reader" do
-    with_stubbed_method(ActiveRecord::Associations::HasManyAssociation, :reader) do
+    assert_not_called_instance_method(ActiveRecord::Associations::HasManyAssociation, :reader) do
       Author.preload(:readonly_comments).first!
-      assert_equal 0, ActiveRecord::Associations::HasManyAssociation
-        .class_variable_get("@@reader_called")
     end
   end
 
diff --git a/activerecord/test/cases/associations/has_many_associations_test.rb b/activerecord/test/cases/associations/has_many_associations_test.rb
index cd6b56fbde..4f4cffbbe3 100644
--- a/activerecord/test/cases/associations/has_many_associations_test.rb
+++ b/activerecord/test/cases/associations/has_many_associations_test.rb
@@ -1,7 +1,6 @@
 # frozen_string_literal: true
 
 require "cases/helper"
-require "support/method_stub"
 require "models/developer"
 require "models/computer"
 require "models/project"
@@ -113,8 +112,6 @@ def test_blank_custom_primary_key_on_new_record_should_not_run_queries
 end
 
 class HasManyAssociationsTest < ActiveRecord::TestCase
-  include MethodStub
-
   fixtures :accounts, :categories, :companies, :developers, :projects,
            :developers_projects, :topics, :authors, :author_addresses, :comments,
            :posts, :readers, :taggings, :cars, :jobs, :tags,
@@ -2137,7 +2134,7 @@ def test_creating_using_primary_key
   end
 
   def test_defining_has_many_association_with_delete_all_dependency_lazily_evaluates_target_class
-    with_stubbed_method(
+    assert_not_called_instance_method(
       ActiveRecord::Reflection::AssociationReflection,
       :class_name
     ) do
@@ -2153,7 +2150,7 @@ class DeleteAllModel < ActiveRecord::Base
   end
 
   def test_defining_has_many_association_with_nullify_dependency_lazily_evaluates_target_class
-    with_stubbed_method(
+    assert_not_called_instance_method(
       ActiveRecord::Reflection::AssociationReflection,
       :class_name
     ) do
diff --git a/activerecord/test/cases/tasks/database_tasks_test.rb b/activerecord/test/cases/tasks/database_tasks_test.rb
index 56acb3ad8c..d8bf3e8968 100644
--- a/activerecord/test/cases/tasks/database_tasks_test.rb
+++ b/activerecord/test/cases/tasks/database_tasks_test.rb
@@ -1,7 +1,6 @@
 # frozen_string_literal: true
 
 require "cases/helper"
-require "support/method_stub"
 require "active_record/tasks/database_tasks"
 
 module ActiveRecord
@@ -46,13 +45,11 @@ def with_stubbed_new
   }
 
   class DatabaseTasksUtilsTask < ActiveRecord::TestCase
-    include MethodStub
-
     def test_raises_an_error_when_called_with_protected_environment
       protected_environments = ActiveRecord::Base.protected_environments
       current_env            = ActiveRecord::Base.connection.migration_context.current_environment
 
-      with_stubbed_method(ActiveRecord::MigrationContext, :current_version, 1) do
+      assert_called_instance_method(ActiveRecord::MigrationContext, :current_version, times: 6, returns: 1) do
         assert_not_includes protected_environments, current_env
         # Assert no error
         ActiveRecord::Tasks::DatabaseTasks.check_protected_environments!
@@ -70,7 +67,7 @@ def test_raises_an_error_when_called_with_protected_environment
     def test_raises_an_error_when_called_with_protected_environment_which_name_is_a_symbol
       protected_environments = ActiveRecord::Base.protected_environments
       current_env            = ActiveRecord::Base.connection.migration_context.current_environment
-      with_stubbed_method(ActiveRecord::MigrationContext, :current_version, 1) do
+      assert_called_instance_method(ActiveRecord::MigrationContext, :current_version, times: 6, returns: 1) do
         assert_not_includes protected_environments, current_env
         # Assert no error
         ActiveRecord::Tasks::DatabaseTasks.check_protected_environments!
@@ -86,7 +83,7 @@ def test_raises_an_error_when_called_with_protected_environment_which_name_is_a_
 
     def test_raises_an_error_if_no_migrations_have_been_made
       ActiveRecord::InternalMetadata.stub(:table_exists?, false) do
-        with_stubbed_method(ActiveRecord::MigrationContext, :current_version, 1) do
+        assert_called_instance_method(ActiveRecord::MigrationContext, :current_version, returns: 1) do
           assert_raise(ActiveRecord::NoEnvironmentInSchemaError) do
             ActiveRecord::Tasks::DatabaseTasks.check_protected_environments!
           end
diff --git a/activerecord/test/cases/tasks/sqlite_rake_test.rb b/activerecord/test/cases/tasks/sqlite_rake_test.rb
index 49d84fb34d..70dd0bf3d1 100644
--- a/activerecord/test/cases/tasks/sqlite_rake_test.rb
+++ b/activerecord/test/cases/tasks/sqlite_rake_test.rb
@@ -1,15 +1,12 @@
 # frozen_string_literal: true
 
 require "cases/helper"
-require "support/method_stub"
 require "active_record/tasks/database_tasks"
 require "pathname"
 
 if current_adapter?(:SQLite3Adapter)
   module ActiveRecord
     class SqliteDBCreateTest < ActiveRecord::TestCase
-      include MethodStub
-
       def setup
         @database      = "db_create.sqlite3"
         @configuration = {
@@ -50,10 +47,8 @@ def test_db_create_when_file_exists
 
       def test_db_create_with_file_does_nothing
         File.stub(:exist?, true) do
-          with_stubbed_method(ActiveRecord::Base, :establish_connection, nil) do
+          assert_not_called_instance_method(ActiveRecord::Base, :establish_connection) do
             ActiveRecord::Tasks::DatabaseTasks.create @configuration, "/rails/root"
-            assert_equal 0,
-              ActiveRecord::Base.class_variable_get("@@establish_connection_called")
           end
         end
       end
diff --git a/activerecord/test/support/method_stub.rb b/activerecord/test/support/method_stub.rb
deleted file mode 100644
index 72e7ba9158..0000000000
--- a/activerecord/test/support/method_stub.rb
+++ /dev/null
@@ -1,53 +0,0 @@
-# frozen_string_literal: true
-
-module MethodStub
-  # FIXME: Are all these tests for method calls (and not called)
-  # really necessary?
-  private
-
-    def with_stubbed_method(object, method_name, return_value = nil)
-      stub_method(object, method_name, return_value)
-
-      yield
-
-    ensure
-      unstub_method(object, method_name)
-    end
-
-    # TODO: `alias_method` is private in Ruby 2.4, so once the minimum
-    # version is 2.5 there'll be no need to resort to `module_eval`.
-
-    def stub_method(object, method_name, return_value = nil)
-      base_var_name = method_name.to_s.gsub(/[?!]/, "")
-
-      object.class_eval(<<-EOF, __FILE__, __LINE__ + 1)
-        @@#{base_var_name}_called = 0
-
-        define_method("stubbed_#{method_name}") do |*args|
-          @@#{base_var_name}_called += 1
-
-          return_value
-        end
-
-        if method_defined?(method_name)
-          alias_method("original_#{method_name}", method_name)
-        end
-        alias_method(method_name, "stubbed_#{method_name}")
-      EOF
-    end
-
-    def unstub_method(object, method_name)
-      base_var_name = method_name.to_s.gsub(/[?!]/, "")
-
-      object.class_eval(<<-EOF, __FILE__, __LINE__ + 1)
-        if method_defined?("original_#{method_name}")
-          alias_method(method_name, "original_#{method_name}")
-          remove_method("original_#{method_name}")
-        else
-          remove_method(method_name)
-        end
-        remove_method("stubbed_#{method_name}")
-        remove_class_variable("@@#{base_var_name}_called")
-      EOF
-    end
-end
diff --git a/activesupport/lib/active_support/testing/method_call_assertions.rb b/activesupport/lib/active_support/testing/method_call_assertions.rb
index c6358002ea..23d89216f5 100644
--- a/activesupport/lib/active_support/testing/method_call_assertions.rb
+++ b/activesupport/lib/active_support/testing/method_call_assertions.rb
@@ -38,6 +38,50 @@ def assert_not_called(object, method_name, message = nil, &block)
         def stub_any_instance(klass, instance: klass.new)
           klass.stub(:new, instance) { yield instance }
         end
+
+        def assert_called_instance_method(klass, method_name, message = nil, times: 1, returns: nil)
+          base_var_name = method_name.to_s.gsub(/[?!]/, "")
+
+          klass.class_eval(<<-EOF, __FILE__, __LINE__ + 1)
+            @@#{base_var_name}_called = 0
+
+            define_method("stubbed_#{method_name}") do |*args|
+              @@#{base_var_name}_called += 1
+
+              returns
+            end
+
+            if method_defined?(method_name)
+              alias_method("original_#{method_name}", method_name)
+            end
+            alias_method(method_name, "stubbed_#{method_name}")
+          EOF
+
+          yield
+
+          times_called = klass.class_variable_get("@@#{base_var_name}_called")
+
+          error = "Expected #{method_name} to be called #{times} times, " \
+            "but was called #{times_called} times"
+          error = "#{message}.\n#{error}" if message
+
+          assert_equal times, times_called, error
+        ensure
+          klass.class_eval(<<-EOF, __FILE__, __LINE__ + 1)
+            if method_defined?("original_#{method_name}")
+              alias_method(method_name, "original_#{method_name}")
+              remove_method("original_#{method_name}")
+            else
+              remove_method(method_name)
+            end
+            remove_method("stubbed_#{method_name}")
+            remove_class_variable("@@#{base_var_name}_called")
+          EOF
+        end
+
+        def assert_not_called_instance_method(klass, method_name, message = nil, &block)
+          assert_called_instance_method(klass, method_name, message, times: 0, &block)
+        end
     end
   end
 end
diff --git a/activesupport/test/testing/method_call_assertions_test.rb b/activesupport/test/testing/method_call_assertions_test.rb
index 0500e47def..aa1211ef4c 100644
--- a/activesupport/test/testing/method_call_assertions_test.rb
+++ b/activesupport/test/testing/method_call_assertions_test.rb
@@ -58,6 +58,45 @@ def test_assert_called_with_message
     assert_match(/dang it.\nExpected increment/, error.message)
   end
 
+  def test_assert_called_instance_method_with_defaults_to_expect_once
+    assert_called_instance_method Level, :increment do
+      @object.increment
+    end
+  end
+
+  def test_assert_called_instance_method_more_than_once
+    assert_called_instance_method(Level, :increment, times: 2) do
+      @object.increment
+      @object.increment
+    end
+  end
+
+  def test_assert_called_instance_method_returns
+    assert_called_instance_method(Level, :increment, returns: 10) do
+      assert_equal 10, @object.increment
+    end
+  end
+
+  def test_assert_called_instance_method_failure
+    error = assert_raises(Minitest::Assertion) do
+      assert_called_instance_method(Level, :increment) do
+        # Call nothing...
+      end
+    end
+
+    assert_equal "Expected increment to be called 1 times, but was called 0 times.\nExpected: 1\n  Actual: 0", error.message
+  end
+
+  def test_assert_called_instance_method_with_message
+    error = assert_raises(Minitest::Assertion) do
+      assert_called_instance_method(Level, :increment, "dang it") do
+        # Call nothing...
+      end
+    end
+
+    assert_match(/dang it.\nExpected increment/, error.message)
+  end
+
   def test_assert_called_with
     assert_called_with(@object, :increment) do
       @object.increment
@@ -107,6 +146,22 @@ def test_assert_not_called_failure
     assert_equal "Expected increment to be called 0 times, but was called 1 times.\nExpected: 0\n  Actual: 1", error.message
   end
 
+  def test_assert_not_called_instance_method
+    assert_not_called_instance_method(Level, :decrement) do
+      @object.increment
+    end
+  end
+
+  def test_assert_not_called_instance_method_failure
+    error = assert_raises(Minitest::Assertion) do
+      assert_not_called_instance_method(Level, :increment) do
+        @object.increment
+      end
+    end
+
+    assert_equal "Expected increment to be called 0 times, but was called 1 times.\nExpected: 0\n  Actual: 1", error.message
+  end
+
   def test_stub_any_instance
     stub_any_instance(Level) do |instance|
       assert_equal instance, Level.new
```

bogdanvlviv commented on activerecord/test/cases/tasks/sqlite_rake_test.rb:
```diff
@@ -47,9 +50,11 @@ def test_db_create_when_file_exists
-          ActiveRecord::Base.expects(:establish_connection).never
-
-          ActiveRecord::Tasks::DatabaseTasks.create @configuration, "/rails/root"
+          with_stubbed_method(ActiveRecord::Base, :establish_connection, nil) do
```
Should be `assert_not_called(ActiveRecord::Base, :establish_connection) {}`

utilum replied:
Quite, fixing.

utilum commented on activerecord/test/support/method_stub.rb:
```diff
@@ -0,0 +1,53 @@
+
+module MethodStub
+  # FIXME: Are all these tests for method calls (and not called)
+  # really necessary?
```
It is not well formulated. How about:

```
  # FIXME: Replace tests for method calls (and not called), especially
  # where the method stubbed is not defined for the object, with simpler
  # tests.
``` 

?

utilum replied:
Pushed the above for now.

utilum commented:
Thank you for all the work (and play) @bogdanvlviv. I like the cleanliness of adding assertions to `ActiveSupport::Testing` (and testing them).
On the other hand, even though in order to get rid of Mocha we are forced to resort to some embedded mocking and to stubbing methods that aren't defined on the object, I think we want to discourage it in general, and try to simplify it where it does exist. Adding it to `ActiveSupport` facilitates permissive testing.
It's a discussion worth having. Is it necessary for this PR ?

bogdanvlviv commented on activerecord/test/support/method_stub.rb:
```diff
@@ -0,0 +1,54 @@
+module MethodStub
+  # FIXME: Replace tests for method calls (and not called), especially
+  # where the method stubbed is not defined for the object, with simpler
+  # tests.
```
Could you provide any example we should fix?

bogdanvlviv commented on activerecord/test/support/method_stub.rb:
```diff
@@ -0,0 +1,54 @@
+    end
+
+    # TODO: `alias_method` is private in Ruby 2.4, so once the minimum
+    # version is 2.5 there'll be no need to resort to `module_eval`.
```
Could you elaborate how it will prevent using of `module_eval` here?
Using `module_eval` seems fine to me here.

utilum replied:
In `RUBY_VERSION > "2.5"` `alias_method` is public and we can avoid resorting to `module_eval` like so:

```
    def stub_method(object, method_name, return_value = nil)
      base_var_name = method_name.to_s.gsub(/[?!]/, "")

      object.class_variable_set("@@#{base_var_name}_called", 0)

      object.define_method("stubbed_#{method_name}") do |*args|
        object.class_variable_set(
          "@@#{base_var_name}_called",
          object.class_variable_get("@@#{base_var_name}_called") + 1
        )

        return_value
      end

      if object.method_defined?(method_name)
        object.alias_method("original_#{method_name}", method_name)
      end
      object.alias_method(method_name, "stubbed_#{method_name}")
    end

    def unstub_method(object, method_name)
      base_var_name = method_name.to_s.gsub(/[?!]/, "")

      if object.method_defined?("original_#{method_name}")
        object.alias_method(method_name, "original_#{method_name}")
        object.remove_method("original_#{method_name}")
      else
        object.remove_method(method_name)
      end
      object.remove_method("stubbed_#{method_name}")
      object.remove_class_variable("@@#{base_var_name}_called")
    end
```

bogdanvlviv replied:
It technically it would be the same, so I'm not sure whether it worth to add this `TODO`
@utilum I've updated my diff (https://github.com/rails/rails/pull/33162#issuecomment-407213833) by adding more tests for these methods and improved the methods.

utilum replied:
It would affect the same modifications to `object`, but it's easier to debug. Also it would be evaluated and checked in `ruby -cw`, as well as validated by `rubocop`.

bogdanvlviv reviewed (commented):
https://github.com/rails/rails/pull/33162#discussion_r204404451
> And then, I like that we need to explicitly assert the expectation.

Yes, It is exactly we need - explicitly assert an expectation.

I want to elaborate why I lean to `assert_not_called_instance_method` and `assert_called_instance_method`:

1) We only need to stub an instance method when it is at least call once, otherwise, there is no reason to stub this method, except we want to test that it method shouldn't be called.

2) `assert_not_called_instance_method`/`assert_called_instance_method` explains that it stubs instace method not method of the passed object, example:
```diff
   # CollectionProxy#reader is expensive, so the preloader avoids calling it.
   test "preloading has_many_through association avoids calling association.reader" do
-    with_stubbed_method(ActiveRecord::Associations::HasManyAssociation, :reader) do
+    assert_not_called_instance_method(ActiveRecord::Associations::HasManyAssociation, :reader) do
       Author.preload(:readonly_comments).first!
-      assert_equal 0, ActiveRecord::Associations::HasManyAssociation
-        .class_variable_get("@@reader_called")
     end
   end
```
since `with_stubbed_method` is confusing about it, See example https://github.com/rails/rails/pull/33162#discussion_r204568565
where I expect from ` with_stubbed_method(ActiveRecord::Base, :establish_connection, nil) {}`
stubs method of `ActiveRecord::Base` but it stubs instance method of this class, so for that we have `assert_not_called` since it similar to `ActiveRecord::Base.expects(:establish_connection).never` and would be consistent to have `assert_not_called_instance_method` for `ActiveRecord::Reflection::AssociationReflection.any_instance.expects(:class_name).never`
3) I think it would be useful for other rails components, so defining that in `activesupport/lib/active_support/testing/method_call_assertions.rb` allows it
4)
https://github.com/rails/rails/pull/33162#issuecomment-407262377
> It's a discussion worth having. Is it necessary for this PR?

I think it would be fine to add by this PR since It more emphasizes the reason of adding these methods and would keep git history clean.

Note that the diff https://github.com/rails/rails/pull/33162#issuecomment-407213833 improves the methods and adds test cases that should work.

utilum commented:
> It's a discussion worth having. Is it necessary for this PR?

>> I think it would be fine to add by this PR since It more emphasizes the reason of adding these methods and would keep git history clean.

It does indeed touch the rationale for this PR, so I'll elaborate, and perhaps something like this should  be added to its description:

> Test doubles are best avoided or kept to a bare minimum; this includes mocking and stubbing. The same holds for testing private methods, undefined methods, and even method calls. I speculate this is one reason that `ActiveSupport::Testing::MethodCalls` is not in the API: We do not encourage use of it.

> Mocha, with its permissive `stub_everything`, `any_instance.stub` and easy stubbing of methods not defined for an object, facilitates, some say promotes, bad testing practices and thereby bad coding practices. Bad practices tend to propagate.

> Before (d4ea114bd611c608482375af94f49e2d54889202):
> ```
> $ cat $(git grep -lE "\.(expects|stubs)\b") | grep -cE "\.(expects|stubs)\b"
> 180
> ```
> Collaborative work on this PR has exposed 49 calls to Mocha methods that plainly served no purpose and could be just be deleted. It also exposed imprecise assertions that required correcting for replicating tests with Minitest's intentionally strict, narrow stubbing and mocking. It also exposed a few instances of unnecessary testing for method calls with `assert_called_with`, where a simple test for value could provide a

> 6/180 calls to Mocha methods proved too resistant to such replication. Until a better solution is found, we stub/mock them with meta-programming.

It seems to me that moving `MethodStub` into `ActiveSupport::Testing::MethodCallAssertions` goes against the direction of this PR by inviting more hackish mocking and stubbing in Rails tests, as well as providing the methods to application developers.

kaspth commented:
Nice! We should definitely test that stub.

I found `assert_called_instance_method` confusing when paired with `assert_called`. How about `assert_called_on_instance_of` instead?

I'm not too kean on all the class_eval though, perhaps a prepended module could work?

```ruby
def assert_called_on_instance_of(klass, method_name, message = nil, times: 1, returns: nil)
  mod = Module.new
  klass.prepend mod

  times_called = 0
  mod.define_method(method_name) { times_called += 1; returns }

  yield

  error = "Expected #{method_name} to be called #{times} times, but was called #{times_called} times"
  error = "#{message}.\n#{error}" if message

  assert_equal times, times_called, error
ensure
  mod.undef_method method_name
end
```

kaspth commented:
Yes, the module prepend would probably get unwieldy and perhaps a memory hog, but for our lightweight internal use case, the increased readability far outweighs the costs. If it works 😄

bogdanvlviv commented:
@kaspth I've run tests that are in diff(https://github.com/rails/rails/pull/33162#issuecomment-407213833) with suggestions in https://github.com/rails/rails/pull/33162#issuecomment-408616175. It doesn't work.

kaspth commented:
@bogdanvlviv what does it fail with? Could we make it work with some fiddling? I'd be happy to look into it, but it probably won't be before this weekend. 😊

bogdanvlviv commented:
@kaspth I've taken look into it:
The first issue was related to using of `undef_method`, we should use `remove_method` in our case, see info about these methods in Ruby docs http://ruby-doc.org/core-2.5.1/Module.html#method-i-undef_method.
The second issue related to method definition - we probably should use something like `define_method(method_name) {|*|}` instead of `define_method(method_name) {}`.

But I have some doubt about this implementation since it doesn't remove a module from ancestors after execution of `assert_called_on_instance_of`/`assert_not_called_on_instance_of`
Example:
```ruby
3.times do
  assert_not_called_on_instance_of(Level, :increment) do
  end
end
Level.ancestors #=> [#<Module:0x000055fbc736bd20>, #<Module:0x000055fbc736be60>, #<Module:0x000055fbc6d29818>, Level, Object, ActiveSupport::Tryable, Kernel, BasicObject]
````

Maybe would be better to go with `class_eval` as it is in https://github.com/rails/rails/pull/33162#issuecomment-407213833?

bogdanvlviv commented:
@kaspth What about

```ruby
def assert_called_on_instance_of(klass, method_name, message = nil, times: 1, returns: nil)
  aliased_method, punctuation = method_name.to_s.sub(/([?!=])$/, ""), $1
  with_method = "#{aliased_method}_with_stub#{punctuation}"
  without_method = "#{aliased_method}_without_stub#{punctuation}"

  times_called = 0

  klass.send(:define_method, with_method) do |*|
    times_called += 1

    returns
  end

  klass.send(:alias_method, without_method, method_name)
  klass.send(:alias_method, method_name, with_method)

  case
  when klass.protected_method_defined?(without_method)
    klass.send(:protected, method_name)
  when klass.private_method_defined?(without_method)
    klass.send(:private, method_name)
  end

  yield

  error = "Expected #{method_name} to be called #{times} times, but was called #{times_called} times"
  error = "#{message}.\n#{error}" if message

  assert_equal times, times_called, error
ensure
  klass.send(:alias_method, method_name, without_method)
  klass.send(:remove_method, with_method)
  klass.send(:remove_method, without_method)
end

def assert_not_called_on_instance_of(klass, method_name, message = nil, &block)
  assert_called_on_instance_of(klass, method_name, message, times: 0, &block)
end
```
It is the similar implementation to https://github.com/rails/rails/pull/33325, and https://github.com/rails/rails/pull/33162#discussion_r204683633.

I also updated the diff in https://github.com/rails/rails/pull/33162#issuecomment-407213833.

kaspth commented:
@bogdanvlviv I like that better, though I don't see the need for the private/protected finagling.

I also don't see the problem with the module prepend. It's an internal method for sparing use. It's much better that it's fewer lines of easier to understand code.

So, why bother unprepending the module? I could shift on that if there were specific failures because of it when you tested @bogdanvlviv?

> 🍬 collaborating with you guys.

Feeling's mutual! ❤️

utilum commented:
The second commit, 6b2f1c6757dd5b62ba434813d9b1f76268752ea4 in this PR is now rebased and includes the tested stubbing in `ActiveSupport::Testing::MethodCallAssertions`.

For the reasons give just above, it does not use a prepended module. @kaspth do you think we can do the same and have an easier to read code by prepending a module?

I've also added a couple of tests to verfiy that we can stub an method that does not exist, and that said introduced method does does not stay defined for the klass after the stubbing op.

utilum commented:
Here's how a prepended module may look:

```
        def assert_called_on_instance_of(klass, method_name, message = nil, times: 1, returns: nil)
          mod = Module.new
          klass.prepend mod

          times_called = 0

          if klass.respond_to?(method_name)
            mod.send(:alias_method, "original_#{method_name}", method_name)
          end

          mod.send(:define_method, method_name) do |*|
            times_called += 1

            returns
          end

          yield

          error = "Expected #{method_name} to be called #{times} times, but was called #{times_called} times"
          error = "#{message}.\n#{error}" if message

          assert_equal times, times_called, error
        ensure
          if mod.method_defined?("original_#{method_name}")
            mod.send(:alias_method, method_name, "original_#{method_name}")
            mod.send(:remove_method, "original_#{method_name}")
          else
            mod.send(:remove_method, method_name)
          end
        end
```
It's a wee longer. Is it easier on the eye?
What am I missing?

kaspth commented on activesupport/lib/active_support/testing/method_call_assertions.rb:
```diff
@@ -35,6 +35,53 @@ def assert_not_called(object, method_name, message = nil, &block)
+            message,
+            times: 0,
+            &block
+          )
```
Let's just keep these arguments on one line, here and elsewhere. Same for the method definition.

utilum replied:
Fixed.

kaspth reviewed (commented):
All right, got around to seeing why my prepended module version didn't work. Thanks guys!

As another crazy attempt I tried to see if I could override this https://github.com/seattlerb/minitest/blob/e6bc4485730403faff6966c1671cf5de72b2d233/lib/minitest/mock.rb#L215 somehow via a wrapper class. Because then we could just use `assert_called`.

Let's mimick minitests aliasing and undef at the least. I think the method defined check in `ensure` is too much.

Pretty close now! ❤️

bogdanvlviv commented:
Isn't https://github.com/rails/rails/pull/33162#issuecomment-409125627 simple enough? We use the same implementation in `deprecate_methods`, see #33325.

kaspth commented:
@bogdanvlviv sure, we can go with that. I'd cut the protected/private stuff and use prefixes, so we don't have to care about the punctuation though.

utilum commented on activesupport/lib/active_support/testing/method_call_assertions.rb:
```diff
@@ -35,6 +35,53 @@ def assert_not_called(object, method_name, message = nil, &block)
 
+        # TODO: No need to resort to #send once support for Ruby 2.4 is
+        # dropped. May even use prepended module. See:
+        # https://github.com/rails/rails/pull/33162#issuecomment-408616175 .
```
There is no need to keep method calls that were only introduced to support a version of Ruby once support for that version is dropped. Unnecessary method calls are not a good idea either for performance code clarity.
I have not checked the other places, but it may be a good idea to check when the time comes.

utilum commented on activesupport/lib/active_support/testing/method_call_assertions.rb:
```diff
@@ -35,6 +35,35 @@ def assert_not_called(object, method_name, message = nil, &block)
+        ensure
+          klass.send(:alias_method, method_name, "original_#{method_name}")
+          klass.send(:undef_method, "original_#{method_name}")
+          klass.send(:undef_method, "stubbed_#{method_name}")
```
Why would we want Ruby to "still search superclasses and mixed-in modules for a possible receiver" of `"original_#{method_name}"` or `"original_#{method_name}"`?

bogdanvlviv replied:
Maybe we shouldn't since using of `undef_method` caused an issue in https://github.com/rails/rails/pull/33162#issuecomment-408616175, https://github.com/rails/rails/pull/33162#issuecomment-409002664

utilum replied:
I don't see that it causes any issue.

bogdanvlviv reviewed (approved):
Excellent! Thanks for working on it.

utilum commented:
I'm not happy with losing the check for `method_defined?`.
Now, if we try to stub a method that's not defined, we get a confusing error message.

```
def test_assert_called_on_instance_of_introduced_method_failure
  assert_called_on_instance_of Level, :introduced_method do
    @object.introduced_method
  end
end
```

output:
```
Error:
MethodCallAssertionsTest#test_assert_called_on_instance_of_introduced_method_failure:
NameError: undefined method `original_introduced_method' for class `MethodCallAssertionsTest::Level'
    /home/u/code/rails/activesupport/lib/active_support/testing/method_call_assertions.rb:59:in `alias_method'
    /home/u/code/rails/activesupport/lib/active_support/testing/method_call_assertions.rb:59:in `ensure in assert_called_on_instance_of'
    /home/u/code/rails/activesupport/lib/active_support/testing/method_call_assertions.rb:61:in `assert_called_on_instance_of'
    /home/u/code/rails/activesupport/test/testing/method_call_assertions_test.rb:165:in `test_assert_called_on_instance_of_introduced_method_failure'
```

Don't we care about this case?

utilum commented:
We could add something like `raise(NameError, "undefined method `#{method_name}' for class #{klass}") unless klass.method_defined?("original_#{method_name}")` to the `ensure` block to make this clear.

bogdanvlviv commented:
Since this only for internal use only, I think it is fine as it is.

utilum commented:
@kaspth, just to be sure, are you fine with the [odd error message](https://github.com/rails/rails/pull/33162#issuecomment-413152431) we'd now return if trying to stub a method that does not exist?

kaspth commented:
@utilum yeah, it's a method for internal use. Let's see if we get hit by it 😊

vbrazo commented:
Good job @utilum
